functions/src/api/lib/validation.ts

```diff
@@ -0,0 +1,134 @@
+import Koa from 'koa';
+import {isEmpty} from 'ramda';
+import * as yup from 'yup';
+
+export class RequestValidationError extends Error {
+  constructor(message: string, validationErrors: yup.ValidationError) {
+    super(message);
+    this.validationError = validationErrors;
+  }
+
+  validationError: yup.ValidationError;
+}
+
+export class ResponseValidationError extends Error {
+  constructor(message: string, validationError: yup.ValidationError) {
+    super(message);
+    this.validationError = validationError;
+  }
+
+  validationError: yup.ValidationError;
+}
+
+type DefaultSchema = yup.AnySchema<unknown, unknown, unknown, ''>;
+interface ValidatedState<
+  TBody extends yup.Schema = DefaultSchema,
+  TQuery extends yup.Schema = DefaultSchema,
+  TResponse extends yup.Schema = DefaultSchema,
+> extends Koa.DefaultState {
+  body: yup.Asserts<TBody>;
+  query: yup.Asserts<TQuery>;
+  response: yup.Asserts<TResponse>;
+}
+
+type Validator<
+  TBody extends yup.Schema,
+  TQuery extends yup.Schema,
+  TResponse extends yup.Schema,
+> = {
+  body: TBody;
+  query: TQuery;
+  response: TResponse;
+};
+
+type ValidatorOptions<
+  TBody extends yup.Schema = DefaultSchema,
+  TQuery extends yup.Schema = DefaultSchema,
+  TResponse extends yup.Schema = DefaultSchema,
+> = {
+  [k in keyof Partial<
+    Validator<TBody, TQuery, TResponse>
+  >]: yup.ValidateOptions;
+};
+
+const defaultOptions: ValidatorOptions = {
+  body: {
+    stripUnknown: true,
+  },
+  query: {
+    stripUnknown: true,
+  },
+  response: {
+    stripUnknown: true,
+  },
+};
+
+const validation =
+  <
+    TBody extends yup.Schema,
+    TQuery extends yup.Schema,
+    TResponse extends yup.Schema,
+  >(
+    validator: Partial<Validator<TBody, TQuery, TResponse>>,
+    options: ValidatorOptions = defaultOptions,
+  ): Koa.Middleware<ValidatedState<TBody, TQuery, TResponse>> =>
+  async (ctx: Koa.Context, next: Koa.Next) => {
+    try {
+      if (validator.body) {
+        ctx.state.body = await validator.body.validate(
```

Comment:
No possibility to update `ctx.body` and `ctx.query` directly here, since it's the default place to pick data from?

Reply:
No, the Koa types didn't seem to approve of that, will check once more

Reply:
Last time I checked body is an object. But query is some kind of prototype with helpers.

Reply:
guess this is the reason koa-yup-validate is using the state object for this